Add getter methods for skolems in API

ajreynol opened the pull request:
This exports SkolemFunId to the API, and adds it to the Java, Python APIs, and adds unit tests.

It does some minor refactoring to remove unnecessary ids and to move abstract value to an internal skolem id.

Followup PRs will add methods for constructing/parsing skolems. I will also rename `SkolemFunId` -> `SkolemId` throughout the code base after this PR is merged.

abdoo8080 commented on include/cvc5/cvc5_skolem_id.h:
```diff
@@ -0,0 +1,333 @@
+{
+  /** The skolem is not exported */
+  EVALUE(INTERNAL),
+  
```
```suggestion
```

abdoo8080 commented on include/cvc5/cvc5_skolem_id.h:
```diff
@@ -0,0 +1,333 @@
+size_t cvc5_skolem_id_hash(Cvc5SkolemFunId rule);
+
+#else
+/**
+ * Converts a proof rule to a string. Note: This function is also used in
+ * `safe_print()`. Changing this function name or signature will result in
+ * `safe_print()` printing "<unsupported>" instead of the proper strings for
+ * the enum values.
+ *
+ * @param id The proof rule
+ * @return The name of the proof rule
+ */
+const char* toString(SkolemFunId id);
```
Use the new `std::to_string` convention in #10361.

ajreynol replied:
Done. Note we still require `toString` for `safe_print`. I've discussed this with @aniemetz  and we will refactor to eliminate the need for `const char * toString` methods throughout.

abdoo8080 commented on include/cvc5/cvc5_skolem_id.h:
```diff
@@ -0,0 +1,333 @@
+  /** abstract value for a term t */
+  EVALUE(ABSTRACT_VALUE),
+  //================================================= Unknown rule
+  EVALUE(NONE),
```
Maybe call it `LAST` for consistency...

ajreynol replied:
We actually need `NONE`, it is the expected return value when something is not a skolem and you ask its id.

We could add `LAST` as well, although I would do this separately when we determine what the overall policy for naming enums should be (note that `ProofRule` also does not use `LAST`).

abdoo8080 commented on src/expr/skolem_manager.cpp:
```diff
@@ -39,77 +39,6 @@ struct UnpurifiedFormAttributeId
-  out << toString(id);
-  return out;
-}
 
```
```suggestion
```

abdoo8080 commented on src/api/cpp/cvc5.cpp:
```diff
@@ -3625,6 +3625,61 @@ Term Term::getRealAlgebraicNumberUpperBound() const
+  SkolemFunId id;
+  skm->isSkolemFunction(*d_node, id, cacheVal);
+  std::vector<Term> args;
+  if (cacheVal.isNull())
```
```suggestion
  if (!cacheVal.isNull())
```

abdoo8080 reviewed (changes requested):
What happened to `SKOLEM_FUN` kind? I prefer to have it (for pattern matching in Lean) than to call `isSkolem` function.

ajreynol commented:
I will add `SKOLEM_FUN` in another PR.

ajreynol commented:
Note that I refactored the name `MAP_PREIMAGE_INJECTIVE` -> `DISTINCT_ELEMENTS_INJECTIVE` for consistency and removed an unused skolem `STRING_OCCUR_LEN`.

mudathirmahgoub commented on src/proof/alf/alf_node_converter.cpp:
```diff
@@ -721,7 +720,7 @@ bool AlfNodeConverter::isHandledSkolemId(SkolemFunId id)
     case SkolemFunId::BAGS_DEQ_DIFF:
     case SkolemFunId::BAGS_DISTINCT_ELEMENTS:
-    case SkolemFunId::BAGS_MAP_PREIMAGE_INJECTIVE:
+    case SkolemFunId::BAGS_DISTINCT_ELEMENTS_INJECTIVE:
```
This enum is specific for map operator, and it refers to a singleton preimage. The name distinct elements is not accurate here.

ajreynol commented:
@mudathirmahgoub ok, I've reverted the name.  Btw, the reason I had changed this is because the documentation for the skolem was missing a reference (it referred to "the same as above" but was no longer clear based on the names).  I've now copied the documentation for `BAGS_MAP_PREIMAGE`  prior to https://github.com/cvc5/cvc5/pull/10499 to `BAGS_MAP_PREIMAGE_INJECTIVE`.  Is this still correct?

mudathirmahgoub commented on include/cvc5/cvc5_skolem_id.h:
```diff
@@ -0,0 +1,759 @@
+ *
+ * This file is part of the cvc5 project.
+ *
+ * Copyright (c) 2009-2023 by the authors listed in the file AUTHORS
```
```suggestion
 * Copyright (c) 2009-2024 by the authors listed in the file AUTHORS
```

mudathirmahgoub commented on include/cvc5/cvc5_skolem_id.h:
```diff
@@ -0,0 +1,759 @@
+ * 
+ * Say that skolem k witnesses the disequality between two arrays A and B
+ * of type ``(Array Int Int)``. Then, k is a term whose skolem identifier is
+ * ``ARRAY_DEQ_DIFF``, skolem indicies are A and B, and whose type is ``Int``.
```
```suggestion
 * ``ARRAY_DEQ_DIFF``, skolem indices are A and B, and whose type is ``Int``.
```

mudathirmahgoub commented on include/cvc5/cvc5_skolem_id.h:
```diff
@@ -0,0 +1,759 @@
+ * to k apart from having the same skolem identifier.
+ * 
+ * In contrast, cvc5 reasons about division-by-zero using a single skolem
+ * function whose identifier is ``DIV_BY_ZERO``. This means its skolem indicies
```
```suggestion
 * function whose identifier is ``DIV_BY_ZERO``. This means its skolem indices
```

mudathirmahgoub commented on include/cvc5/cvc5_skolem_id.h:
```diff
@@ -0,0 +1,759 @@
+   * Difference index for string disequalities, such that k is the witness for
+   * the inference
+   *  (=> (not (= a b)) (not (= (substr a k 1) (substr b k 1))))
+   * where note that `k` may be out of bounds for at most of of a,b.
```
```suggestion
   * where note that `k` may be out of bounds for at most of a,b.
```

mudathirmahgoub commented on include/cvc5/cvc5_skolem_id.h:
```diff
@@ -0,0 +1,759 @@
+  EVALUE(BAGS_MAP_INDEX),
+  /**
+   * An uninterpreted function for bag.map operator:
+   * If the preimage of {y} in A is {uf(1), ..., uf(n)} (see BAGS_DISTINCT_ELEMENTS},
```
```suggestion
   * If bag A is {uf(1), ..., uf(n)} (see BAGS_DISTINCT_ELEMENTS},
```

mudathirmahgoub commented on include/cvc5/cvc5_skolem_id.h:
```diff
@@ -0,0 +1,759 @@
+   * A, and it is mapped to y by f.
+   *
+   * - Number of skolem indices: ``2``
+   *   - ``1:`` a map term of the form ``(set.map f A)`` where A is ``(Set E)``
```
```suggestion
   *   - ``1:`` a map term of the form ``(set.map f A)`` where A of type ``(Set E)``
```

mudathirmahgoub commented on include/cvc5/cvc5_skolem_id.h:
```diff
@@ -0,0 +1,759 @@
+#else
+
+/**
+ * Writes a proof rule name to a stream.
```
```suggestion
 * Writes a skolem id to a stream.
```

mudathirmahgoub commented on include/cvc5/cvc5_skolem_id.h:
```diff
@@ -0,0 +1,759 @@
+ * Writes a proof rule name to a stream.
+ *
+ * @param out The stream to write to
+ * @param id The proof rule to write to the stream
```
```suggestion
 * @param id The skolem id to write to the stream
```

mudathirmahgoub commented on include/cvc5/cvc5_skolem_id.h:
```diff
@@ -0,0 +1,759 @@
+{
+  /**
+   * Hashes a SkolemFunId to a size_t.
+   * @param rule The proof rule.
```
```suggestion
   * @param id The skolem id.
```

mudathirmahgoub commented on include/cvc5/cvc5_skolem_id.h:
```diff
@@ -0,0 +1,759 @@
+   * @param rule The proof rule.
+   * @return The hash value.
+   */
+  size_t operator()(cvc5::SkolemFunId rule) const;
```
```suggestion
  size_t operator()(cvc5::SkolemFunId id) const;
```

mudathirmahgoub commented on src/api/cpp/cvc5_skolem_id.cpp:
```diff
@@ -0,0 +1,46 @@
+ *
+ * This file is part of the cvc5 project.
+ *
+ * Copyright (c) 2009-2023 by the authors listed in the file AUTHORS
```
```suggestion
 * Copyright (c) 2009-2024 by the authors listed in the file AUTHORS
```

mudathirmahgoub commented on src/api/cpp/cvc5_skolem_id.cpp:
```diff
@@ -0,0 +1,46 @@
+ * directory for licensing information.
+ * ****************************************************************************
+ *
+ * Implementation of proof rule.
```
```suggestion
 * Implementation of skolem id.
```

mudathirmahgoub commented on src/api/java/CMakeLists.txt:
```diff
@@ -58,6 +58,24 @@ add_custom_command(
 add_custom_target(generate-java-proofrules DEPENDS ${JAVA_PROOF_RULE_FILE})
 
+# Generate SkolemFunId.java
+set(JAVA_SKOLEM_FUN_ID_FILE
```
This file needs to be added to the list `JAVA_FILES` defined here
https://github.com/cvc5/cvc5/blob/main/src/api/java/CMakeLists.txt#L91. 
Otherwise, the jar file would not include this class. 
The CI should capture this, which implies the unit tests somehow finds the class outside the jar file. 
We need to update the CI in future PR to use the jar file instead of compiled files.

ajreynol replied:
Should we add `ProofRule.java` here too?

mudathirmahgoub replied:
Yes. According to the documentation here 
https://cmake.org/cmake/help/latest/module/UseJava.html. Currently the jar file includes ProofRule.java may be because it is referenced by other classes that are in the jar file. To be safe, we need to add it to the jar file as well.

mudathirmahgoub commented on src/api/java/io/github/cvc5/Term.java:
```diff
@@ -803,6 +803,47 @@ public Term getRealAlgebraicNumberUpperBound()
+  /**
+   * @api.note Asserts isSkolem().
+   * @api.note This method is experimental and may change in future versions.
+   * @return The proof rule used by the root step of the proof.
```
```suggestion
   * @return The skolem identifier of this term.
```

mudathirmahgoub commented on src/api/java/io/github/cvc5/Term.java:
```diff
@@ -803,6 +803,47 @@ public Term getRealAlgebraicNumberUpperBound()
+  private native boolean isSkolem(long pointer);
+
+  /**
+   * @api.note Asserts isSkolem().
```
```suggestion
   * Get skolem identifier of this term.
   * @api.note Asserts isSkolem().
```

mudathirmahgoub commented on src/expr/skolem_manager.cpp:
```diff
@@ -230,7 +158,20 @@ Node SkolemManager::mkSkolemFunctionTyped(SkolemFunId id,
       std::stringstream ss;
-      ss << "@" << id;
+      if (id == SkolemFunId::INTERNAL)
+      {
```
I don't know why this case needs special treatment. Can you add documentation for it?

ajreynol replied:
Done.

mudathirmahgoub commented on src/printer/enum_to_string.h:
```diff
@@ -0,0 +1,45 @@
+ * Utilities for printing API enum values.
+ *
+ * In particular, we require `toString(...)` to be defined for any identifier
+ * that may appear in statistics, for safe printing (src/util/safe_print.h).
```
So far only skolem id are included in this file. So statistics here seems out of place. I guess all enum printers would be added here.

ajreynol replied:
Yes, I discussed this with @aniemetz offline. We decided to add this file and gradually migrate the API enum values that require both `std::to_string` and `toString` to this header.

mudathirmahgoub commented on test/unit/api/cpp/term_black.cpp:
```diff
@@ -1208,6 +1208,15 @@ TEST_F(TestApiBlackTerm, getRealAlgebraicNumber)
+TEST_F(TestApiBlackTerm, getSkolem)
+{
+  // ordinary variables are not skolems
+  Term x = d_tm.mkConst(d_tm.getIntegerSort(), "x");
```
Currently we don't allow users to create skolems, do we? This is why this test seems trivial.

ajreynol replied:
Yes, creation of skolems will be done in a followup PR, which will make it easy to add new unit tests.

mudathirmahgoub reviewed (approved):
LGTM
